File: packages/react-native/ios/DashXRequestInterceptor.swift
import Foundation
import Alamofire

class DashXRequestInterceptor: RequestInterceptor {
    private let retryLimit: Int = 3
    private let exponentialBackoffScale: Double = 0.5
    private let jitterFactor: Double = 0.1
    private let exponentialBackoffBase: Int = 2
    
    private func getTimeDelay(for retryCount: Int) -> TimeInterval {
        let randomFactor: Double = 1 + (1 - Double.random(in: 0..<1) * 2) * jitterFactor
        var batteryState: UIDevice.BatteryState { UIDevice.current.batteryState }
        switch batteryState {
        case .charging:
            return (pow(Double(exponentialBackoffBase), Double(retryCount)) * exponentialBackoffScale) * randomFactor
        default:
            return (pow(Double(exponentialBackoffBase), Double(retryCount + 1)) * exponentialBackoffScale) * randomFactor
        }
    }
    
    private func checkStatusCode(_ request: Request) -> Bool {
        let retryableStatusCodes = [408, 500, 502, 503, 504]
        if let statusCode = request.response?.statusCode,
            retryableStatusCodes.contains(statusCode) {
            return true
        } else {
            return false
        }
    }
    
    private func checkMethod(_ request: Request) -> Bool {
        if let httpMethod = request.request?.method,
            httpMethod == .post {
            return true
        } else {
            return false
        }
    }
    
    func retry(_ request: Request, for session: Session, dueTo error: Error, completion: @escaping (RetryResult) -> Void) {
        if request.retryCount < retryLimit,
            checkMethod(request),
            checkStatusCode(request) {
            let timeDelay = getTimeDelay(for: request.retryCount)
            completion(.retryWithDelay(timeDelay))
        } else {
            completion(.doNotRetry)
        }
    }
}

File: packages/react-native/android/src/main/java/com/dashx/sdk/DashXRequestInterceptor.kt
package com.dashx.sdk

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.os.BatteryManager
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import kotlin.math.pow

class DashXRequestInterceptor(
    private val exponentialBackoffScale: Double = .5,
    private val exponentialBackoffBase: Double = 2.0,
    private val retryLimit: Int = 3,
    private val jitterFactor: Double = .1
) : Interceptor {
    val context = DashXClient.instance.reactApplicationContext

    private fun shouldRetry(request: Request, response: Response): Boolean {
        val connectivityManager = context?.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = connectivityManager.activeNetworkInfo
        val retryableStatusCodes = listOf<Int>(408, 500, 502, 503, 504)

        return (networkInfo.isConnected && !networkInfo.isRoaming)
            && request.method == "POST"
            && retryableStatusCodes.contains(response.code)
    }

    private fun getTimeDelay(retryCount: Int): Long {
        val randomFactor = 1 + (1 - Math.random() * 2) * jitterFactor

        val batteryStatus: Intent? = IntentFilter(Intent.ACTION_BATTERY_CHANGED).let { it ->
            context?.registerReceiver(null, it)
        }
        val extraStatus = batteryStatus?.getIntExtra(BatteryManager.EXTRA_STATUS, -1)

        return if (extraStatus == BatteryManager.BATTERY_STATUS_CHARGING) {
            ((exponentialBackoffBase.pow(retryCount) * exponentialBackoffScale) * randomFactor).toLong()
        } else {
            ((exponentialBackoffBase.pow(retryCount + 1) * exponentialBackoffScale) * randomFactor).toLong()
        }
    }

    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        var response = chain.proceed(request)
        var retryCount = 1

        while (retryLimit >= retryCount && shouldRetry(request, response)) {
            Thread.sleep(getTimeDelay(retryCount))
            retryCount++
            response = chain.proceed(request)
        }

        return response
    }
}
